connectors/sources/sharepoint.py

```diff
@@ -0,0 +1,715 @@
+            document[elasticsearch_field] = (
+                item[item_type][sharepoint_field]
+                if document_type in [DRIVE_ITEM]
+                else item[sharepoint_field]
+            )
+
+        return document
+
+    async def ping(self):
+        """Verify the connection with Sharepoint"""
+
+        self._validate_configuration()
+
+        if not self.ssl_disabled:
+            self.ssl_ctx = self._ssl_context(certificate=self.certificate)
+
+        self.session = await self._generate_session()
+        try:
+            await anext(
+                self._api_call(
+                    url_name=PING,
+                    site_collections=self.site_collections[0],
+                    host_url=self.host_url,
+                )
+            )
+            logger.info("Successfully connected to the Sharepoint")
+        except Exception:
+            logger.exception("Error while connecting to the Sharepoint")
+            raise
+
+    async def get_content(
+        self, document, file_relative_url, site_url, timestamp=None, doit=None
+    ):
+        """Get content of list items and drive items
+
+        Args:
+            document (dictionary): Modified document.
+            file_relative_url (str): Relative url of file
+            site_url (str): Site path of sharepoint
+            timestamp (timestamp, optional): Timestamp of item last modified. Defaults to None.
+            doit (boolean, optional): Boolean value for whether to get content or not. Defaults to None.
+
+        Returns:
+            dictionary: Content document with id, timestamp & text.
+        """
+
+        if not (self.enable_content_extraction and doit and document["size"]):
+            return
+
+        document_size = int(document["size"])
+        if document_size > FILE_SIZE_LIMIT:
+            logger.warning(
+                f"File size {document_size} of file {document['title']} is larger than {FILE_SIZE_LIMIT} bytes. Discarding file content"
+            )
+            return
+
+        source_file_name = ""
+
+        async with NamedTemporaryFile(mode="wb", delete=False) as async_buffer:
+            async for response in self._api_call(
+                url_name=ATTACHMENT,
+                host_url=self.host_url,
+                value=site_url,
+                file_relative_url=file_relative_url,
+            ):
+                async for data in response.content.iter_chunked(CHUNK_SIZE):
+                    await async_buffer.write(data)
+            source_file_name = async_buffer.name
+
+        await asyncio.to_thread(
+            convert_to_b64,
+            source=source_file_name,
+        )
+        async with aiofiles.open(file=source_file_name, mode="r") as target_file:
+            # base64 on macOS will add a EOL, so we strip() here
+            attachment_content = (await target_file.read()).strip()
+        await remove(source_file_name)
```

Comment:
Is it possible to make it a utility function that you re-use in all connectors?

Reply:
No it is not possible to make it a utility function because we are calling `_api_call` method on line 48, which can create issue(unclosed session).

Reply:
I think you can delegate it to the function by providing the function that returns chunks as one of parameters for this function, something like this (name is purely for example):

Does it make sense? cc @tarekziade 

```python
async def download_and_stream(async_batch_generator):
    async with NamedTemporaryFile(mode="wb", delete=False) as async_buffer:
            async for data in async_batch_generator():
                await async_buffer.write(data)
            source_file_name = async_buffer.name

        await asyncio.to_thread(
            convert_to_b64,
            source=source_file_name,
        )
        async with aiofiles.open(file=source_file_name, mode="r") as target_file:
            # base64 on macOS will add a EOL, so we strip() here
            attachment_content = (await target_file.read()).strip()
        await remove(source_file_name)
```